lib/Sema/CSSimplify.cpp

```diff
@@ -1410,6 +1410,22 @@ ConstraintSystem::matchTypes(Type type1, Type type2, ConstraintKind kind,
     case ConstraintKind::ArgumentConversion:
     case ConstraintKind::OperatorArgumentTupleConversion:
     case ConstraintKind::OperatorArgumentConversion:
+      if (type1->is<InOutType>() && typeVar2 &&
+          typeVar2->getImpl().canBeInOut()) {
+        // Upgrade type2 to an InOutType
+        auto inOutType1 = type1->getAs<InOutType>();
+        
+        unsigned typeVar3Options = typeVar2->getImpl().getOptions();
+        typeVar3Options &= ~TVO_CanBeInOut;
+        auto typeVar3 = createTypeVariable(getConstraintLocator(locator),
+                                           typeVar3Options);
+        auto inOutType3 = InOutType::get(typeVar3);
+        
+        addConstraint(ConstraintKind::Equal, typeVar2, inOutType3, locator);
+        return matchTypes(inOutType1->getObjectType(), typeVar3, kind, flags,
```

Comment:
Even if that were the case this is going to bind the extra typevar anyway in the course of solving the system.  I think the new TV Option can be removed in favor of a materializable constraint and a fixed binding.